### ISSUE TITLE
Unable to create new session using a custom profile

### ISSUE THREAD

**Issue by soosrol** [mozilla/geckodriver#1058]
## System
* Version: 0.19.1
* Platform: Windows 10
* Firefox: 57.0
* Selenium: 3.7.1 (Java)

## Testcase

1. Set up Selenium hub + node
2. Prepare Firefox profile on the node (optionally install an extension)
3. Create a browser on the grid with passing in the args to load the profile from local file system:

```
FirefoxOptions fo = new FirefoxOptions();
fo.addArguments("-profile", "./firefoxprofile");
DesiredCapabilities c = DesiredCapabilities.firefox();
c.setCapability(FirefoxOptions.FIREFOX_OPTIONS, fo);
RemoteWebDriver wd = new RemoteWebDriver(hubUrl, c);
```
### Expected result

1. Browser opened
2. Profile is loaded from the node's local file system
3. Session is created
4. Browser can be interacted with using the RemoteWebDriver object from the tests

### Actual result
1. Browser opened
2. Profile is loaded from the node's local file system
3. RemoteWebDriver.createSession() never returns, session is not created

### Notes

- The Firefox profile was created on the same node, with the same version of Firefox
- There is only one extension installed, but I tried it without that, same result
- Starting Firefox through selenium grid works as expected without `fo.addArguments("-profile", "./firefoxprofile");`
- Firefox does not hang when started, it can be interacted with manually.
- Launching Firefox with `firefox.exe -profile "path/to/profile"` also works.

## Stacktrace
```
10:50:23.300 INFO - Got a request to create a new session: Capabilities {acceptInsecureCerts: true, androidServerPort: 7100, appInstall: true, browserName: firefox, moz:firefoxOptions: {args: [-profile, ./firefoxprofile], log: {level: trace}, prefs: {}}, serverPort: 5769, uuid: f39b5203-64ca-4b0f-b4c2-b0c..., version: }
10:50:23.300 INFO - Trying to create a new session on test slot {firefox_binary=C:\Program Files\Mozilla Firefox\firefox.exe, seleniumProtocol=WebDriver, browserName=firefox, maxInstances=1, platform=WIN10}
11:00:23.363 INFO - Error forwarding the new session new session request for webdriver should contain a location header or an 'application/json;charset=UTF-8' response body with the session ID.
org.openqa.grid.common.exception.GridException: new session request for webdriver should contain a location header or an 'application/json;charset=UTF-8' response body with the session ID.
        at org.openqa.grid.internal.TestSession.updateHubIfNewWebDriverSession(TestSession.java:426)
        at org.openqa.grid.internal.TestSession.forward(TestSession.java:248)
        at org.openqa.grid.web.servlet.handler.RequestHandler.forwardNewSessionRequestAndUpdateRegistry(RequestHandler.java:90)
        at org.openqa.grid.web.servlet.handler.RequestHandler.process(RequestHandler.java:113)
        at org.openqa.grid.web.servlet.DriverServlet.process(DriverServlet.java:84)
        at org.openqa.grid.web.servlet.DriverServlet.doPost(DriverServlet.java:68)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.seleniumhq.jetty9.servlet.ServletHolder.handle(ServletHolder.java:841)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doHandle(ServletHandler.java:535)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doScope(ServletHandler.java:473)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doScope(SessionHandler.java:1564)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.seleniumhq.jetty9.server.Server.handle(Server.java:564)
        at org.seleniumhq.jetty9.server.HttpChannel.handle(HttpChannel.java:317)
        at org.seleniumhq.jetty9.server.HttpConnection.onFillable(HttpConnection.java:251)
        at org.seleniumhq.jetty9.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
        at org.seleniumhq.jetty9.io.FillInterest.fillable(FillInterest.java:110)
        at org.seleniumhq.jetty9.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
        at org.seleniumhq.jetty9.util.thread.Invocable.invokePreferred(Invocable.java:128)
        at org.seleniumhq.jetty9.util.thread.Invocable$InvocableExecutor.invoke(Invocable.java:222)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:294)
        at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:199)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:590)
        at java.lang.Thread.run(Unknown Source)
```

## Trace-level log
```
C:\SeleniumGrid>java -Dwebdriver.chrome.driver="./chromedriver.exe" -Dwebdriver.gecko.driver="./geckodriver.exe" -jar selenium-server-standalone.jar -role node -hub http://lpselgrid3001.noclab.local:4444/grid/register/ -nodeConfig nodeconfig.json
10:49:54.363 INFO - Selenium build info: version: '3.7.1', revision: '8a0099a'
10:49:54.363 INFO - Launching a Selenium Grid node
2017-11-15 10:49:55.050:INFO::main: Logging initialized @1133ms to org.seleniumhq.jetty9.util.log.StdErrLog
10:49:55.113 INFO - Driver class not found: com.opera.core.systems.OperaDriver
10:49:55.128 INFO - Driver provider class org.openqa.selenium.safari.SafariDriver registration is skipped:
 registration capabilities Capabilities {browserName: safari, platform: MAC, version: } does not match the current platform WIN10
10:49:55.175 INFO - Using the passthrough mode handler
2017-11-15 10:49:55.206:INFO:osjs.Server:main: jetty-9.4.5.v20170502
2017-11-15 10:49:55.222:WARN:osjs.SecurityHandler:main: ServletContext@o.s.j.s.ServletContextHandler@1c9d8e2{/,null,STARTING} has uncovered http methods for path: /
2017-11-15 10:49:55.238:INFO:osjsh.ContextHandler:main: Started o.s.j.s.ServletContextHandler@1c9d8e2{/,null,AVAILABLE}
2017-11-15 10:49:55.253:INFO:osjs.AbstractConnector:main: Started ServerConnector@13eba10{HTTP/1.1,[http/1.1]}{0.0.0.0:5555}
2017-11-15 10:49:55.253:INFO:osjs.Server:main: Started @1331ms
10:49:55.253 INFO - Selenium Grid node is up and ready to register to the hub
10:49:55.253 INFO - Starting auto registration thread. Will try to register every 5000 ms.
10:49:55.253 INFO - Registering the node to the hub: http://lpselgrid3001.noclab.local:4444/grid/register
10:49:55.363 INFO - The node is registered to the hub and ready to use
2017-11-15 10:50:23.331:INFO:osjshC.ROOT:qtp25671582-16: org.openqa.selenium.remote.server.WebDriverServlet-691363: Initialising WebDriverServlet
10:50:23.347 INFO - Found handler: org.openqa.selenium.remote.server.commandhandler.BeginSession@b3462b
10:50:23.347 INFO - /session: Executing POST on /session (handler: BeginSession)
10:50:23.425 INFO - Capabilities are: Capabilities {acceptInsecureCerts: true, androidServerPort: 7100, appInstall: true, browserName: firefox, firefox_binary: C:\Program Files\Mozilla Fi..., moz:firefoxOptions: {args: [-profile, ./firefoxprofile], binary: C:\Program Files\Mozilla Fi..., log: {level: trace}, prefs: {}}, serverPort: 5769, uuid: f39b5203-64ca-4b0f-b4c2-b0c..., version: }
10:50:23.441 INFO - Capabilities {acceptInsecureCerts: true, androidServerPort: 7100, appInstall: true, browserName: firefox, firefox_binary: C:\Program Files\Mozilla Fi..., moz:firefoxOptions: {args: [-profile, ./firefoxprofile], binary: C:\Program Files\Mozilla Fi..., log: {level: trace}, prefs: {}}, serverPort: 5769, uuid: f39b5203-64ca-4b0f-b4c2-b0c..., version: } matched class org.openqa.selenium.remote.server.ServicedSession$Factory (provider: org.openqa.selenium.firefox.GeckoDriverService)
1510771823519   geckodriver     INFO    geckodriver 0.19.1
1510771823519   geckodriver     INFO    Listening on 127.0.0.1:35467
1510771824050   mozrunner::runner       INFO    Running command: "C:\\Program Files\\Mozilla Firefox\\firefox.exe" "-marionette" "-profile" "./firefoxprofile"
1510771826113   geckodriver::marionette TRACE     connection attempt 0/600
1510771828253   geckodriver::marionette TRACE     connection attempt 1/600
1510771830410   geckodriver::marionette TRACE     connection attempt 2/600
1510771832550   geckodriver::marionette TRACE     connection attempt 3/600
[...]
```

**Comment by whimboo** [comment]
So when you log into the node, can you see that Firefox is getting started at all? I would expect at least the Firefox log to be present in the trace log. I assume you followed https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/TraceLogs.html?

**Comment by andreastt** [comment]
I am also surprised to not find any stdout output from Firefox after the `Running command: …` log line. If you see `connection attempt` going up to 600 and then timing out, it means geckodriver is either unable to launch Firefox or that it is unable to connect to the Marionette remote protocol in the Firefox that gets started.

**Comment by 2mf** [comment]
I'm having the same issue with Selenium 3.7.0 and geckodriver 0.19.1.
However in my case firefox is starting and geckodriver can't connect to firefox.
When I use the geckodriver 0.18.0 it is working again.

**Comment by soosrol** [comment]
@whimboo, @andreastt: 
Yes, if I log into the node, I can see Firefox starting up, and it's responsive, I can interact with it manually. But somehow Geckodriver can't connect. If I don't try to load the profile, it usually connects on the 2nd-3rd try. 
I'll see if I can squeeze out more log detailed logs, but I'm already using `{"moz:firefoxOptions": {"log": {"level": "trace"}}}`. Is there a separate Firefox log file somewhere? Currently I'm getting these logs from the hub/node's console out.

**Comment by soosrol** [comment]
@2mf:
Firefox starts for me too, but geckodriver can't connect. I'll check if it works with v0.18. Are you also trying to load a profile with command line args, or just starting FF normally?

**Comment by whimboo** [comment]
@soosrol, does Firefox in those failing cases show the robot in the location bar? Also have a look at the browser console (tools menu) if there are some errors reported for Marionette.

Regarding the logs, usually all output should (for the time being) end up in the same log file. But I don't know anything about the remove driver to give a qualified answer in how it works.

**Comment by soosrol** [comment]
@whimboo, yes, it has the robot in the address bar. Also checked the browser console, and as far as I can tell, it doesn't have any relevant errors. See attached screenshots.

![screen shot 2017-11-17 at 9 27 07 am](https://user-images.githubusercontent.com/16192335/32953492-5b1eb17e-cb7e-11e7-9dd9-44e2b27df454.png)
![screen shot 2017-11-17 at 9 59 24 am](https://user-images.githubusercontent.com/16192335/32953497-5e6c96c0-cb7e-11e7-9b05-26850894cd20.png)

**Comment by 2mf** [comment]
@soosrol I'm loading a profile in a few cases but most cases do not use profiles at all.
What I'm doing is running Firefox headless or within Xvfb inside a docker container.

**Comment by whimboo** [comment]
@soosrol interesting! It reminds me on a case we occasionally see in our automated tests but which is so seldom that I wasn't able to hit it so far.

Is that reproducible for you all the time? Does it also happen when you are not creating a new profile but let geckodriver do it? I would like to know what the minimal testcase is for you.

Also I would like to see a real trace level log. Would be fantastic if you can find out where this is stored on the hub.

Btw. is that a new issue you are facing with the 57 release, or did it already exist with 56?

**Comment by soosrol** [comment]
@whimboo I can reproduce it every time. Everything works flawlessly if I don't specify a profile to load. This has been the same ever since I first tried it 2-3 Firefox versions ago and since Selenium 3.4 and geckodriver 17 or so...

What's really strange, is that **I only encounter this issue if I pass in the command line arg for loading the profile from local drive.** If I generate/load the profile within my tests by using
```
FirefoxProfile profile = new FirefoxProfile(new File("path/to/profile"));
FirefoxOptions options = new FirefoxOptions();
options.setProfile(profile);
capabilities.setCapability(FirefoxOptions.FIREFOX_OPTIONS, options);
```
the profile gets loaded and and the session is started without an issue.

You may ask why I'm not doing it this way, and the answer is simple: I am running my tests on Selenium grid on multiple threads and this would require me to have selenium transfer the profile upon every browser start/restart. My custom profile contains a browser extension, that is ~20MB in size. Having a profile of this size transferred by selenium grid is not just very slow, but also very fragile, as after 2-3 tries I always get Java heap space exception, regardless of the allowed memory in JVM settings.

**Comment by whimboo** [comment]
Thank you for the update @soosrol. It looks great to have a reduced testcase now. One thing you could help us with is, can you please try if you can get the same issue reproduced when not running the test on the hub but locally? Does that produce the same failure?

Otherwise, @andreastt do you have an idea? Without having trace logs for that problem, I'm meanwhile out of ideas.

**Comment by soosrol** [comment]
@whimboo: I've tried running the same comparison (starting with/without the profile arg) on my local machine and I got the exact same results. I think this rules out most of the environmental causes, as my local machine is a Mac, unlike my grid that runs on Windows VMs.

See attached logs for both sessions:
Without profile arg: [good_ff.log](https://github.com/mozilla/geckodriver/files/1511517/good_ff.log)
With profile arg: [bad_ff.log](https://github.com/mozilla/geckodriver/files/1511518/bad_ff.log)

_Selenium version: 3.7.1
Geckodriver version: 19.1
Firefox version: 57.0 (64-bit)_

**Comment by whimboo** [comment]
Amazing to see. It means we are getting closer! Given those results I took the opportunity and had a look myself. Here the results...

When trying to use an already existent profile the `profiles` key in `moz:firefoxOptions` cannot be used. It indeed only supports base64 encoded profile data. But as [our documentation](https://github.com/mozilla/geckodriver/#firefox-capabilities) says: `To have geckodriver pick up an existing profile on the filesystem, please set the args field to {"args": ["-profile", "/path/to/your/profile"]}`

So I tried to use `args` and passed-in the path to a specific profile directory. Firefox launches, by using this profile but we are not able to make any connection to Marionette even it is running. After the connection failure the test quits without killing the Firefox instance. I will continue diving deeper into that...

@soosrol can you please also test the `args` key. Do you get the correct profile loaded? Or do you have the same connection problems?

**Comment by whimboo** [comment]
The problem with not being able to make a connection to Marionette for the given profile path is, that by default Marionette uses port 2828, but geckodriver takes a random free port. Right now I do not see a way to allow pass-in a fixed port to geckodriver when not starting it manually. Checking the code I can only see that the command line argument `--marionette-port` gets evaluated.

So whether we need a capability under `moz:firefoxOptions` to let users specify a port for Marionette, or geckodriver would have to update the `prefs.js` file first with the wanted port before starting Firefox. @andreastt what do you think?

**Comment by soosrol** [comment]
@whimboo: Yes, the `args` works, that is how I am loading the profile. The correct profile gets loaded both on Windows/Mac, regardless if I do it with the grid or locally.

In my opinion you may have just found the root cause with the marionette port!

When I don't specify the profile and let it be generated for the session, I can see on `about:config`  that the port was set via `prefs.js`. However, when loading my profile with `-profile`, it shows as default just like you said.

At this point I'm trying to see if I can somehow force the geckodriver `--marionette-port` command line parameter from either the code or some config.

**Comment by whimboo** [comment]
I will file a bug soon so that it can be fixed.

As workaround you could start geckodriver manually by passing in `--marionette-port 2828` or whatever you have in the profile setup. Sadly there is no other way possible right now.

**Comment by soosrol** [comment]
Yes, that worked. I ended up pointing `webdriver.gecko.driver` to a shell script like this:

```
#!/bin/bash
path/to/geckodriver "$@" --marionette-port 2828
```

**Comment by whimboo** [comment]
[Bug 1421766](https://bugzilla.mozilla.org/show_bug.cgi?id=1421766) is now taking care of it.

Thank you for reporting! We really wonder why no-one else has seen this yet.

**Comment by whimboo** [comment]
And good to know that there is a workaround at least.

**Comment by stewart-r** [comment]
Any chance someone could clarify the workaround steps for this? I'm using the selenium standalone-firefox-debug docker container and selenium remote webdriver on .NET

I tried specifying `--marionette-port 2828`  in `options.AddArgument` and can see from the container logs that it is being passed to the CLI:

    1512734718509   mozrunner::runner       INFO    Running command: "/usr/bin/firefox" "-marionette" "--profile /home/seluser/.mozilla/firefox/5vgqsnyv.default/" "--marionette-port 2828"

but I still have the same behaviour. I'm not at all sure how to check which port is in the profile setup? :-/ I'm not too familiar with selenium or geckodriver and feel like I am missing the point.

**Comment by stewart-r** [comment]
Ah i understand now... In case anyone else was struggling in the same way as I was the `--marionette-port` command line option is applied when starting geckodriver not starting firefox.

So workaround steps are:

1. rename /usr/bin/geckodriver to /usr/bin/geckodriver2
2. use @soosrol 's script to point to the renamed executable file and save this script in the original /usr/bin/geckodriver location

Many thanks for the workaround, guys! All back up and running now :-)

**Comment by phandinhloc** [comment]
I got the same issue. Specifying the marionette port makes it very difficult to write the tests that start multiple firefox profile at the same time.

**Comment by usmanpasha** [comment]
FAILED CONFIGURATION: @BeforeClass beforeMethod
org.openqa.selenium.WebDriverException: org.apache.http.conn.HttpHostConnectException: Connect to localhost:32703 [localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
Build info: version: '3.7.0', revision: '2321c73', time: '2017-11-02T22:22:35.584Z'
System info: host: 'MV-IMAGE-E17-8.local', ip: '2001:f40:902:a86d:4920:6e2e:35df:85c6', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.12.6', java.version: '1.8.0_131'
Driver info: driver.version: FirefoxDriver

**Comment by oneumyvakin** [comment]
For applying workaround on Windows I've used following Go code: 
```
package main

import (
	"os"
	"os/exec"
	"fmt"
)

func main() {
	var args []string = os.Args[1:]
	args = append(args, "--marionette-port")
	args = append(args, "2828")

	cmd := exec.Command(
		"geckodriver2.exe",
		args...,
	)
	fmt.Println(args)
	err := cmd.Run()
	if err != nil {
		fmt.Printf("failed to execute '%s' '%s': %s\r\n", cmd.Path, cmd.Args, err)
		return
	}
}
```

**Comment by whimboo** [comment]
Well, the actual workaround for now in case you are using the Python bindings will look like the following:
```
opts = Options()
opts.binary = binary
opts.add_argument("-profile")
opts.add_argument(self.profile)

capabilities = DesiredCapabilities.FIREFOX.copy()
driver = webdriver.Firefox(options=opts,
                           capabilities=capabilities,
                           service_args=["--marionette-port", "2828"])
```
Other bindings should also expose the given options and the `services_args` argument.

**Comment by alkayata** [comment]
I've found same problem with selenium 3.12, firefox 60.1 and geckodriver 0.20.1
The workaround worked fine.

**Comment by abalfour** [comment]
@stewart-r Thank you for the tips! 

Unfortunately I'm still struggling to get this working. Is the shell script being saved as a .sh, or is there a way to convert this to an .exe instead? When I run my program it seems that the Selenium driver is still looking for geckodriver.exe at the original location and isn't invoking the script that I've left in its place. I assume if I were to just leave the original geckodriver.exe in the original location that Selenium would just run that rather than invoking the .sh.

Also, does adding `Ooptions.AddArgument($"--marionette-port 2828");` not do anything at all, or is that also necessary for the fix to work?

**Comment by whimboo** [comment]
@abalfour please see my [last comment](https://github.com/mozilla/geckodriver/issues/1058#issuecomment-390256212). It explains it in Python but should be similar for other Selenium bindings.

**Comment by alkayata** [comment]
@abalfour. It seems you are using windows. You can tell selenium to
look for the geckodriver executable like this: "java
-Dwebdriver.gecko.driver='path to your geckodriver executable' -jar
./selenium-server-standalone-3.12.0.jar". Don't forget to use the -D
flag before the -jar or it doesn't work.
 Besides, I think if you want to use a shell script in Windows, you
need to use a .bat file. So taking that into account, to start the
selenium server you should use something like that: "java
-Dwebdriver.gecko.driver='path to your geckodriver executable.BAT' -jar
./selenium-server-standalone-3.12.0.jar"
 Last, you need to adapt your shell script to Windows. The one provided
works with bash, a command interpreter for unix, mainly linux. What it
does is to spawn a geckodriver instance passing the parameters given to
the script to it. For example, if I call 'geckodriver.sh -watermellon',
it will in turn call 'geckodriver -watermellon --marionette-port 2828'
. The '@' sign is the equivalente for 'all arguments passed to the
script'. So you need to find the way to do the same in windows. I don't
think it will be hard.
 I hope it works

**Comment by hornschorsch** [comment]
Is there any chance that this bug will be corrected some time?

**Comment by umarch06** [comment]
@alkayata can you please suggest me solution for C#, here is what I am doing but browser doesn't do anything just open and then no action is performed. 

```
Proxy pro = new Proxy();
                pro.HttpProxy = PROXY;
                pro.FtpProxy = PROXY;
                pro.SslProxy = PROXY;
                //pro.SocksProxy = PROXY;

                FirefoxOptions firefoxOptions = new FirefoxOptions();
                //firefoxOptions.Proxy = pro;
                firefoxOptions.AddArgument("-profile " + Configuration.BrowserProfilePath);
                firefoxOptions.AddArgument("--marionette-port " + "2828");

                PropertiesCollection.Driver = new FirefoxDriver(firefoxOptions);
```

**Comment by whimboo** [comment]
@umarch06 so Firefox doesn't have an argument `marionette-port`. You have to add it to the call of the `FirefoxDriver` constructor instead. But I don't know how the C# implementation handles that exactly.

**Comment by umarch06** [comment]
Can you share sample for any other language? I'll highly appreciate it.

**Comment by whimboo** [comment]
I cannot help here given that I didn't work with C# yet. And as such I don't know the driver at all. But you can check the API docs at https://seleniumhq.github.io/selenium/docs/api/dotnet/.

**Comment by pmaddi2** [comment]
Even after following work around steps, still able to encounter marionette connection issue.

Geckodriver20 or 18
Firefox 52
Selenium 3.13 or 3.4
Java

With custom or default profile path arguments, always listening port is 2828 and not entering into next step "ie.:- driver.get(URL); and traces are showing these:---
"Marionette	INFO	Listening on port 2828"
"geckodriver::marionette	TRACE	  connection attempt 6/600"
"geckodriver::marionette	TRACE	  connection attempt 6/600"

If there is no profile argument provided, Marionette is taking some random available port, and Marionette connection is successful and able to execute further URL navigation step.

Code Used with or without Marionette and with Custom or default profile path arguments is:-
//ffoptions.addArguments("--marionette-port 2828");
	  ffoptions.setAcceptInsecureCerts(true);
	  ffoptions.setLogLevel(FirefoxDriverLogLevel.TRACE);
	  //ffoptions.addArguments("--binary C:\\Program Files (x86)\\Mozilla Firefox\\firefox.exe");
	  ffoptions.addArguments("-profile");
	  //ffoptions.addArguments("C:\\defaultprofilepath\\Optum.default");
	  ffoptions.addArguments(customProfilePath);
    driver = new FirefoxDriver(ffoptions);
driver.get(URL);

Also Tried with below code able to navigate further steps (i.e.:- App URL, etc..,), but unable to replace/append custom(temp) & default profile cookies with successful authenticated cookies.
  ffoptions.setAcceptInsecureCerts(true);
  ffoptions.setLogLevel(FirefoxDriverLogLevel.TRACE);
   driver = new FirefoxDriver(ffoptions);
   driver.get(url);

Could anyone please advise me solution for navigating geckodriver with custom profile. Please let me know if you need any further details in-order to provide solution

**Comment by andreastt** [comment]
Firefox 52 is not supported by geckodriver. Can we _please_ only add comments that are actually relevant to the issue.

**Comment by pmaddi2** [comment]
Actually I tried with version 62 nightly build, still able to reproduce, showing connection refused, not going to navigate url step...

**Comment by whimboo** [comment]
@pmaddi2 as it looks like the Java bindings of Selenium don't allow you to specify a port for Marionette. As least I cannot find anything in the Java API docs. So you will have to wait until this issue has been fixed.

**Comment by pmaddi2** [comment]
Thank you Henrik,

I'll wait for the fix.
In parallel if you get any alternate solution for profile loading. (like extension to override Marionette listening port before driver triggers it),etc. then could you please advise me/share in this thread. It is happening only with Custom profile Version 52.8.1 & 55. 
And Geckodriver 21 or 20.1 with firefox 62/63 nightly build is starting by taking some random free port(not 2828), but showing connection refused 500 internal server error. I retried by restarting machine and & clearing cache. still no luck.

**Comment by pmaddi2** [comment]
Thank you @whimboo

I’ll wait

**Comment by mnemyx** [comment]
@whimboo Thanks for the clarification in Python. 

My scenario isn't exactly one-to-one; I'm using a remote Selenium Grid on a Windows Server with the profile on a local machine, not on the node itself. So if it helps anyone with a similar setup:

1.) I followed @soosrol's lead but the batch file ends up as:
`C:\bin\geckodriver.exe %* --marionette-port 2828`

2.) Node init on the grid:
`java -Dwebdriver.firefox.driver=C:\bin\geckodriver.bat -jar C:\selenium\selenium-server-standalone-3.13.0.jar -role node -hub http://localhost:4444/grid/register`

3.) And python:
```python
opts = webdriver.FirefoxOptions()
prof = webdriver.FirefoxProfile("/path/to/local/profile")
opts.profile = prof
driver = webdriver.Remote(command_executor="http://remote.hub", options=opts)
```

Thanks again for saving me several hours of frustration!

**Comment by artemklushin** [comment]
@umarch06 For c# you can try next solution:
`FirefoxDriverService ffDriverService = FirefoxDriverService.CreateDefaultService(appPath);
  ffDriverService.BrowserCommunicationPort = 2828;
  IWebDriver driver = new FirefoxDriver(ffDriverService, ffOptions);`

It's works for me with profile in custom directory.

**Comment by xbxolivesupport** [comment]
follow this code 
package demo1;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.remote.DesiredCapabilities;

public class Executer2 {

public static void main(String[] args){
	
	
	DesiredCapabilities capabilities = DesiredCapabilities.firefox();
	capabilities.setCapability("acceptInsecureCerts", true);
	capabilities.setJavascriptEnabled(true);
	capabilities.setCapability("marionette", true);
	WebDriver driver = new FirefoxDriver(capabilities);
	System.setProperty("webdriver.gecko.driver", System.getProperty("user.dir") + "/geckodriver.exe");
		driver.get("http://www.google.com");
		driver.close();
				
	
}
any issue regarding browser get help from this 

https://www.browsertechnicalsupportnumbers.com/blog/fix-mozilla-firefox-gmail-error-code-007/
--

**Comment by cloudinfra** [comment]
The workaround is awesome.  we renamed the original gekodriver-v0.21 to gekodriver_original, and the wapper script to gekodriver-v0.21.  the script used gekodriver_original with the --marionette-port 2828.  Also, the port 2828 is specified in the user.js file inside the profile you are passing.  so if you find a different port number there, you have to accordingly use the correct one. (marionette.port)

**Comment by Deinceps** [comment]
I found that the prefs.js file contains the port on which marionette runs
and on that basis I made my decision on c#

```
FirefoxOptions options = new FirefoxOptions();
options.AddArguments("-profile", profliePath);
var service = FirefoxDriverService.CreateDefaultService();
string port = Regex.Match(
File.ReadAllText(profliePath + "/prefs.js"), 
  "marionette.port\", (\\d+?)" + Regex.Escape(")"))
  .Groups[1]
  .Value;
service.BrowserCommunicationPort = int.Parse(port);
var driver = new FirefoxDriver(service, options);
```

**Comment by bamBrewing** [comment]
### System
    Version: 0.23.0
    Platform: Linux
    Firefox: 60.3esr

I am not able to create a new session using a custom profile.  
- Start geckodriver on port 4444
- Post the following to the session endpoint (https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/Capabilities.html#capabilities-example):
```
{
  "capabilities":{
    "alwaysMatch":{
      "moz:firefoxOptions":{
        "binary":"/bin/firefox",
        "args":[
          "-profile",
          "/path/to/my/profile"
        ]
      }
    }
  }
}
```
- geckodriver will log:
`Running command: "/bin/firefox" "-marionette" "-profile" "/path/to/my/profile" "-foreground" "-no-remote"`
- Firefox will open and the profile is loaded as expected
- After 60 seconds a 500 will be returned from the server with a connection refused message.
- If I remove the alwaysMatch from the capabilities:
```
{
  "capabilities":{
    "moz:firefoxOptions":{
      "binary":"/bin/firefox",
      "args":[
        "-profile",
        "/path/to/my/profile"
      ]
    }
  }
}
```
- geckodriver will log:
`Running command: "/bin/firefox" "-marionette" "-profile" "/tmp/rust_mozprofile.6JBkouREH02c" "-foreground" "-no-remote"`
- Firefox will open, but the profile is not loaded
- The session is created using, and my profile settings are not loaded.

**Comment by parthibd** [comment]
I want to confirm that this issue still persists with latest version of selenium and  geckodriver . Marionette will always listen on port 2828 . Is there a way to temporarily fix this issue ?

**Comment by hornschorsch** [comment]
I have made a patch that fixes the problem by opening the file prefs.js in the profile directory and replacing the value of marionette.port with the actual portnumber.

https://manitu.koblenz-net.de/~horn/export/geckodriver-0.24.0-horn.patch
https://manitu.koblenz-net.de/~horn/export/geckodriver-0.24.0

**Comment by whimboo** [comment]
@hornschorsch that is great to hear. Would you mind to upload your patch for review so that we can work on it to get it into an official release? The related bug is https://bugzilla.mozilla.org/show_bug.cgi?id=1421766. The process for review can be found at https://firefox-source-docs.mozilla.org/testing/marionette/marionette/NewContributors.html. Thanks a lot!

**Comment by hornschorsch** [comment]
Hello Henrik,

Am Mo den 11. März 2019 um 02:20 Uhr schrieb Henrik Skupin:

> @hornschorsch that is great to hear. Would you mind to upload your
> patch for review so that we can work on it to get it into an official
> release? The related bug is
> https://bugzilla.mozilla.org/show_bug.cgi?id=1421766. The process for
> review can be found at
> https://firefox-source-docs.mozilla.org/testing/marionette/marionette/NewContributors.html.

I am very sorry but this procedure is to complicated. I donate this patch
to you. Would you mind to review it, possibly rewrite it (i am not
experienced in programming in rust and possibly this can be done better
or eleganter) and then integrate it into geckodriver if you want?

Thanks a lot!

Kind Regards,
Georg

**Comment by TheDan64** [comment]
I've also run into this issue, pretty much identically to what @bamBrewing described but in Windows and FF 66. FF will open with the specified profile but then geckdriver will not do anything with eventually crashing with a 500 ISE...

**Comment by edmolten** [comment]
I needed to use the profile and also some unsigned extensions, so I used a developer edition of Firefox 66. Even using the workaround, Firefox still created a "dev profile" and was copying the content of my profile to it, but my profile was used anyway, super weird. As my profile is quite big, I didn't want it to be copied every time I used it, so I found that adding an empty file called `ignore-dev-edition-profile` alongside the `profiles.ini` file made Firefox stop copying it.

> Ah i understand now... In case anyone else was struggling in the same way as I was the `--marionette-port` command line option is applied when starting geckodriver not starting firefox.
> 
> So workaround steps are:
> 
> 1. rename /usr/bin/geckodriver to /usr/bin/geckodriver2
> 2. use @soosrol 's script to point to the renamed executable file and save this script in the original /usr/bin/geckodriver location
> 
> Many thanks for the workaround, guys! All back up and running now :-)

**Comment by AlexeiScherbakov** [comment]
This behavior is crazy. Chrome driver is worked as expected. When use XPCOM
directly or with GeckoFX (.Net wrapper) on old Firefox/Waterfox - there is
no crazy profile copying, just folder directory. Why behavior of
Geckodriver canno be the same? Just profile folder, no copying, no other
'shadow magic'?

чт, 2 мая 2019 г., 17:26 Eduardo Ramírez Farias <notifications@github.com>:

**Comment by umarch06** [comment]
@artemklushin I tried your solution and it works, but the problem is that if you will try to run firefox with next profile - it will open the last used profile and that's totally weird. are you experiencing the same issue?

**Comment by lacell75** [comment]
I need this correction asap. I didn't find workaround for my case:  I use 4 instances of firefox with selenium Grid. I can't use a fix marionette port.

**Comment by lacell75** [comment]
Hello, @hornschorsch , your geckodriver patched is for linux or windows version?

**Comment by bolds07** [comment]
STILL DOESN'T WORK

LOOKS LIKE SOMEONE IS PAID TO CLOSE ISSUES WITHOUT REALLY PROVIDING A SOLUTION

**Comment by CashApp2020** [comment]
We are mobile payment service serving on a mission to provide the best fluent and fluid experience in mobile transactions, feel free to contact [cash app customer service assistance](https://www.apponlinesolutions.com/) for your once stop solution .

**Comment by veliakiner** [comment]
@soosrul's excellent workaround inspired me to discover you can do the same thing using the `service args` argument for `webdriver.Firefox`:
`service_args=["--marionette-port", "2828"]`
If you're running multiple browsers at once you'll of course need to tell other browsers to listen on a different port. You can do that by setting ` desired_capability["marionette.port"] = <port of your choice>` in `webdriver.DesiredCapabilities.FIREFOX`

**Comment by kretschmannj** [comment]
I got the -profile option to work as expected by adding --marionette-port .  Here's the firefox section from my nightwatch.conf.js file ...

firefox: {
	desiredCapabilities: {
		browserName: 'firefox',
		alwaysMatch: {'moz:firefoxOptions': {args: ['-profile', 'Firefox_Profile/Nightwatch']}}
	},
	webdriver: {
		start_process: true,
		server_path: (Services.geckodriver ? Services.geckodriver.path : ''),
		'cli_args': ['--marionette-port', '2828']
	}
}
Note that I'm using a relative path to the profile file which I have saved in the Firefox_Profile folder in my project.  Also, the marionette-port number doesn't seem to matter much so I just picked some arbitrary value.

**Comment by tttony** [comment]
> 
> 
> @umarch06 For c# you can try next solution:
> 
> ```cs
> FirefoxDriverService ffDriverService = FirefoxDriverService.CreateDefaultService(<driver path>);
> ffDriverService.BrowserCommunicationPort = 2828;
> IWebDriver driver = new FirefoxDriver(ffDriverService, <instance of FirefoxOptions>);
> ```
> 
> It's works for me with profile in custom directory.

This worked right away!! Thanks

I created a new profile running `firefox.exe -p` then I and setted in the options:

```
FirefoxOptions options = new FirefoxOptions();
options.AddArguments("-profile", @"C:\Users\<user>\AppData\Roaming\Mozilla\Firefox\Profiles\8oh5nwar.selenium"); 
```
There was no need to parse the marionette port in the `prefs.js` file, I think it's because the profle is from actual firefox profile(?) in that file there is no `marionette-port` option

**Comment by sybrandy** [comment]
Hello,

I'm having the same issue trying to use a custom profile using the selenium-side-runner CLI tool.  My current .side.yml looks liek this:

```
capabilities:
        acceptInsecureCerts: true
        browserName: firefox
        moz:firefoxOptions:
                args: ["-profile", "/tmp/profile/"]
                log:
                        level: trace
```

Below is some sample output from my test run:

```
16:42:53.116 INFO [ActiveSessionFactory.apply] - Capabilities are: {
  "acceptInsecureCerts": true,
  "browserName": "firefox",
  "moz:firefoxOptions": {
    "args": [
      "-profile",
      "\u002ftmp\u002fprofile\u002f"
    ],
    "log": {
      "level": "trace"
    }
  }
}
16:42:53.116 INFO [ActiveSessionFactory.lambda$apply$11] - Matched factory org.openqa.selenium.grid.session.remote.ServicedSession$Factory (provider: org.openqa.selenium.firefox.GeckoDriverService)
1594744973172   mozrunner::runner       INFO    Running command: "/usr/bin/firefox" "-marionette" "-profile" "/tmp/profile/" "-foreground" "-no-remote"
1594744973174   geckodriver::marionette DEBUG   Waiting 60s to connect to browser on 127.0.0.1:34696
1594745033269   mozrunner::runner       DEBUG   Killing process 411
1594745033276   webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"timeout","message":"connection refused","stacktrace":""}}
```

Does anyone have an idea as to how I can work around this using the runner?  I'm using version 3.141.59 of the selenium standalone Firefox webdriver image.

**Comment by baptx** [comment]
Hello, I tried to use a custom profile like that but it did not work, did I do something wrong?
Go to a Firefox profile folder and create a ZIP file of a profile with this command: `zip -r profile.zip *`
Create the Base64 with this command: `base64 profile.zip > profile_zip_base64.txt`
Use the profile in the Selenium WebDriver script using Node.js like this:
```
var webdriver = require('selenium-webdriver');
var firefox = require('selenium-webdriver/firefox');
var options = new firefox.Options();
options.addArguments("-profile /tmp/profile_zip_base64.txt");
var capabilities = webdriver.Capabilities.firefox();
var driver = new webdriver.Builder()
var driver = new webdriver.Builder()
	.usingServer("http://localhost:4444")
	.withCapabilities(capabilities)
	.setFirefoxOptions(options)
	.build();
driver.get('https://www.google.com/?hl=en');
```
Then I start GeckoDriver with the command: `geckodriver --marionette-port 2828`
But when I execute the script, it opens the web page in my default Firefox profile. The ZIP profile that I created should also contain a cookie named "test" that I created on google.com but I don't see it in the web console.

**Comment by dncpax** [comment]
> Ah i understand now... In case anyone else was struggling in the same way as I was the `--marionette-port` command line option is applied when starting geckodriver not starting firefox.
> 
> So workaround steps are:
> 
> 1. rename /usr/bin/geckodriver to /usr/bin/geckodriver2
> 2. use @soosrol 's script to point to the renamed executable file and save this script in the original /usr/bin/geckodriver location
> 
> Many thanks for the workaround, guys! All back up and running now :-)

I know it's been a while, but I'm struggling with this bug... and for the life of me I cannot figure out what the workaround is!!? What's this script? Can anyone still help?

**Comment by baptx** [comment]
@dncpax in case it helps, the workaround I used is to create a custom WebExtension and load it with Selenium WebDriver instead of loading a whole Firefox profile: https://github.com/SeleniumHQ/selenium/issues/8693#issuecomment-691498083
The advantage is that it will be more lightweight, you only load what you need. But profile support could be useful also. Without profile support, if you need to save cookies / localStorage / sessionStorage, you would need to to it manually (writing and reading data on a file).

**Comment by dncpax** [comment]
@baptx thanks for replying. My goal is to login automatically so Im using a profile for that which has autoauth extension installed and configured. With this bug I'm stuck. I also use selenium-side-runner to avoid scripting tests. So I'm really stuck... This is such an old bug...

**Comment by baptx** [comment]
@dncpax the workaround I use should work for you also, if you load the WebExtension instead of the profile (and write / read session ID stored in e.g. cookies to a file if necessary).

**Comment by dncpax** [comment]
@baptx that would be wonderful... but I'm working with windows auth... that's the only reason I need a profile and the autoauth extension. Would this work with a webextension?

**Comment by hornschorsch** [comment]
A fix for this problem is above in this thread. No idea why it isn't
implemented.

**Comment by dncpax** [comment]
@hornschorsch same here... I know everyone does what he/she can, but someone posted 10lines of working code, and somehow it wasn't picked up... not to rant or anything...

**Comment by baptx** [comment]
@dncpax What do you mean by windows auth? It is probably possible to do it without profile if you write and read profile data to a file. Where is the 10 lines of working code exactly?

**Comment by dncpax** [comment]
windows integrated authentication...

**Comment by globalcitizen** [comment]
## Verified

This bug has been open for 4 years.

Still seeing this issue on `geckodriver-0.30.0` with `selenium-webdriver-4.0.3` .

- **Same desire** (use a prior session with cookies).
- **Same result** (selenium never gains control of the browser, but otherwise the session works fine).

## Further investigation
 * Only occurs when passing the existing profile with `-P <profilename>`
 * Does not occur when passing the existing profile with `-p </path/to/profile>`

## Fix
 * Pass your profile with `-p </path/to/profile>` instead of `-P <profile_name>`

**Comment by saifahn** [comment]
@globalcitizen Does using `-p` over `-P` actually use the profile correctly? Thanks :)

**Comment by globalcitizen** [comment]
@saifahn Sorry test environment long since destroyed.

**Comment by ToxicFrog** [comment]
> @globalcitizen Does using `-p` over `-P` actually use the profile correctly? Thanks :)

I've experimented with this and it does not appear to -- `-p` and `-P` are aliases for each other, `--profile` has no short equivalent. So when running with `-p path/to/profile/dir`, it looks for a profile with that *name*, doesn't find one, and ends up running in an ephemeral profile instead -- you can verify this by running it in non-headless mode and looking at `about:profiles`.

If you use `-P profile-name`, `-p profile-name`, or `--profile path/to/profile`, it doesn't work.

**Comment by whimboo** [comment]
Can someone who still sees this problem please provide the code that you are using with Selenium? When using a custom profile you will have to make sure to instruct geckodriver to use the appropriate port for Marionette which should be passed to `FirefoxDriver()` (at least for the Python binding) as `service_args`.

**Comment by ToxicFrog** [comment]
@whimboo here's how I'm using it, via etaoin:

```
(wd/firefox
   {:args ["-P" "test.profile"]
    :args-driver ["--log" "trace" "--marionette-port" port]
    :log-stdout "/dev/tty"
    :log-stderr "/dev/tty"
    :size [1920 2160]
    :path-browser firefox-bin
    :locator "css selector"})
```
Note that by "doesn't work" earlier, I don't mean "it doesn't load the profile", I mean "it loads the profile but doesn't autonegotiate a port to use for Marionette, so unless you manually set one in `prefs.js` or `user.js` ahead of time, it'll always pick 2828 and you can only have one instance running at a time".

In the above code, I've constructed an appropriate `user.js` by hand and `port` is passed in via the command line, but this doesn't scale; absent a fix for this, I'm probably going to have to write something that figures out where the profile directory is, selects a port, and writes an appropriate `user.js` before launching geckodriver with `--marionette-port` (hoping that nothing else has stolen that port in the intervening time).

**Comment by whimboo** [comment]
@ToxicFrog thank you for the details. For now I don't have a proper solution yet, but I can say that we are working on it. But please bear with us that it might take some time to finish it. For now here a workaround:

1. Prepare your profile with the preference `marionette.port` set to `0`.
2. Let your automation start Firefox directly without using geckodriver by passing `--marionette --profile %profile_path%` and other arguments if needed
3. Read the file `%profile_path%/MarionetteActivePort` for the port Marionette is using
4. Use geckodriver with `--connect-existing --marionette-port %port%` to connect geckodriver to Marionette

We are aware that this is not optimal but at least a start so that scaling won't be a problem for you anymore. With the above steps each instance of Firefox and Marionette will pick up a free and unique system port.

Ongoing work can be tracked via: https://bugzilla.mozilla.org/show_bug.cgi?id=1240830#c16 and ff. For now only step 1 has been done.

**Comment by whimboo** [comment]
Good news. Recently we landed a patch for geckodriver which by default would allow Marionette to use a system allocated port, and geckodriver then reads that port from the Firefox profile. Anyone with the problem on this issue could you please check [this Nightly build of geckodriver](https://treeherder.mozilla.org/jobs?repo=mozilla-central&searchStr=gd-s&revision=559c2c231f2187c471b1e482e85ad9b96218800f). Just click the green `gd-s` job for your platform, and download the `geckodriver.tar.gz` or `geckodriver.zip` from by clicking the link in the lower pane.

When testing please don't force a Marionette port via the `--marionette-port` argument of geckodriver because that would turn off the new code path. Also you have to make sure to at least use a Firefox 95 release.

Please let us know if it works or not. If it's the latter please attach a trace log. Thanks!

**Comment by ToxicFrog** [comment]
I've been otherwise occupied this week but I'll try that out next time I'm tinkering with this, which should hopefully be soon. Thanks!

**Comment by ToxicFrog** [comment]
No luck.

In short: geckodriver starts up, and starts firefox; firefox does not write the MarionetteActivePort file, and geckodriver spins endlessly waiting for it to appear until its parent process watchdogs and kills it.

The runtime profile directory it's looking for in /run/user does exist.
The profile has a blank `user.js` in the firefox config directory, and the `user.js` in /run/user has `marionette.port` set to 0.

It looks like geckodriver is doing the right thing but FF isn't writing the MarionetteActivePort file; any advice on next steps?

```
1644624687959   geckodriver     INFO    Listening on 127.0.0.1:45335
1644624688072   webdriver::server       DEBUG   -> POST /session {"desiredCapabilities":{"loggingPrefs":{"browser":"ALL"},"moz:firefoxOptions":{"binary":"/nix/store/pjxk10j2cbj9ld1xz78wq1m1jggylcy2-firefox-96.0.3/bin/firefox","args":["-width","1920","-height","2160","--new-window","https://[redacted]/","--headless","-P","ca.ancilla.hangbrain"]}}}
1644624688072   webdriver::command      WARN    You are using deprecated legacy session negotiation patterns (desiredCapabilities/requiredCapabilities), see https://developer.mozilla.org/en-US/docs/Web/WebDriver/Capabilities#Legacy
1644624688072   geckodriver::capabilities       DEBUG   Trying to read firefox version from ini files
1644624688072   geckodriver::capabilities       DEBUG   Trying to read firefox version from binary
1644624688144   geckodriver::capabilities       DEBUG   Found version 96.0.3
1644624688145   mozrunner::runner       INFO    Running command: "/nix/store/pjxk10j2cbj9ld1xz78wq1m1jggylcy2-firefox-96.0.3/bin/firefox" "--marionette" "-width" "1920" "-height" "2160" "--new-window" "https://chat.google.com/" "--headless" "-P" "ca.ancilla.hangbrain" "-no-remote"
1644624688145   geckodriver::marionette DEBUG   Waiting 60s to connect to browser on 127.0.0.1
1644624688145   geckodriver::browser    TRACE   Failed to open /run/user/8509/rust_mozprofileKu4Rax/MarionetteActivePort
1644624688145   geckodriver::marionette TRACE   Retrying in 100ms
*** You are running in headless mode.
1644624688246   geckodriver::browser    TRACE   Failed to open /run/user/8509/rust_mozprofileKu4Rax/MarionetteActivePort
1644624688246   geckodriver::marionette TRACE   Retrying in 100ms
Crash Annotation GraphicsCriticalError: |[0][GFX1-]: glxtest: Unable to open a connection to the X server (t=0.12545) [GFX1-]: glxtest: Unable to open a connection to the X server
Crash Annotation GraphicsCriticalError: |[0][GFX1-]: glxtest: Unable to open a connection to the X server (t=0.12545) |[1][GFX1-]: glxtest: libEGL initialize failed (t=0.125485) [GFX1-]: gl
xtest: libEGL initialize failed
1644624688346   geckodriver::browser    TRACE   Failed to open /run/user/8509/rust_mozprofileKu4Rax/MarionetteActivePort
1644624688346   geckodriver::marionette TRACE   Retrying in 100ms
[...many more such messages redacted...]
Crash Annotation GraphicsCriticalError: |[0][GFX1-]: glxtest: Unable to open a connection to the X server (t=0.12545) |[1][GFX1-]: glxtest: libEGL initialize failed (t=0.125485) |[2][GFX1-]: RenderCompositorSWGL failed mapping default framebuffer, no dt (t=0.790276) [GFX1-]: RenderCompositorSWGL failed mapping default framebuffer, no dt
```
The X11/GL errors show that FF is starting up successfully, and are normal -- it's running headless on a machine with no X server. I can see that Geckodriver and Firefox are both running:

```
$ pgrep -laf gecko
1504017 geckodriver --port 35947
$ pgrep -laf firefox
1517150 /nix/store/pjxk10j2cbj9ld1xz78wq1m1jggylcy2-firefox-96.0.3/bin/.firefox-wrapped --marionette -width 1920 -height 2160 --new-window https://[redacted]/ --headless -P ca.ancilla.hangbrain -no-remote
$ firefox --version                           
Mozilla Firefox 96.0.3
$ geckodriver --version     
geckodriver 0.30.0 (b9121c1a4330 2022-01-27 23:34 +0200)
```

**Comment by whimboo** [comment]
Interesting case. So you are specifying a profile name from the profile manager when using `"-P","ca.ancilla.hangbrain"`. But geckodriver doesn't know about the profile location and as such is checking the wrong directory. Can you please check by using the profile path via `"-profile", "%path_to_profile%"`?

For the `-P` issue I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1755261.

**Comment by ToxicFrog** [comment]
~~I can do that as a one-off, but not in the general case since the mapping from profile name to profile directory isn't straightforward~~
If I use `-CreateProfile "foo path/to/foo"`, can I then subsequently use `-profile path/to/foo`? In that case all the profile names are predictable and I can use this.

**Comment by whimboo** [comment]
@ToxicFrog can you explain why you need `-CreateProfile`? You can start Firefox with `-profile path` to let it populate the folder with a fresh profile. That way it doesn't pollute the profile manager entries. Note that by using the profile manager the path to the profile is platform dependent.

**Comment by ToxicFrog** [comment]
I'd completely forgotten about that -- I think that would work better for my use case anyways. I'll try that out and report back, thanks.

**Comment by ToxicFrog** [comment]
It works! It even works with an existing profile, as long as you don't have `marionette.port` set in it.

**Comment by whimboo** [comment]
> It works! It even works with an existing profile, as long as you don't have `marionette.port` set in it.

This is great to hear! So yes, `marionette.port` should not be set or have a value of `0`. Thanks for the verification.

If there is still a problem related to this feature please file a new issue. Otherwise we will ship this new feature with the next geckodriver 0.31.0 release.

**Comment by whimboo** [comment]
> For the `-P` issue I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1755261.

FYI this bug is now fixed and we marked the usage of `-P` as deprecated. Instead of the profile name the profile path has to be used.

**Comment by MagdelineNg** [comment]
@whimboo Hi, I am referencing this issue to automate the existing Firefox instance by connecting to the marionette.port. According to this issue and many others, it seems like the default port is 2828 for marionette, but I realised the selenium driver connects to the port shown in the image instead (and that marionette.port variable changes every time I relaunch the Firefox browser). Is there a way to fix the marionette port when launching Firefox?

<img width="1256" alt="image" src="https://github.com/mozilla/geckodriver/assets/85625305/7637c24f-ce19-4045-97dc-be77b65e6117">

Current workflow: 
1. Launching Firefox
```
'C:\Program Files\Mozilla Firefox\firefox.exe' - marionette 
'C:\Program Files\Mozilla Firefox\firefox.exe' - marionette -start-debugger-server=2828 #also tried this but marionette.port still does not reflect 2828
```

2. Launching Selenium GeckoDriver
```
driver = webdriver.Firefox(executable_path=GeckoDriverManager().install(), service-args=['--marionette-port', '65493', '--connect-existing'])   #i don't want to hardcode the port, is there a way to get the Firefox marionette.port variable?
print(driver.page_source)
```

**Comment by whimboo** [comment]
First note that `-start-debugger-server` is not an option that controls Marionette. Instead it is related to DevTools. So you can omit that argument.

Regarding connecting to an existing instance of Firefox this is not yet fully supported. Especially not with random ports given that geckodriver cannot read the `MarionetteActivePort` file from the user's Firefox profile. But if you know which profile is used you could read the port yourself and feed it as argument to the Selenium driver.
